Add support for serializing debug_value instructions

usama54321 opened the pull request:
This PR adds support for serialization and deserialization of debug value instructions

usama54321 committed:
Initial commit for serializing/deserializing debug value instructions

usama54321 committed:
Serialize generic environment for all functions when serializing debug
info.

usama54321 committed:
Bugfix. The previous debug scope is not necessarily the function scope

usama54321 committed:
Do not emit shadow copies in deserialized functions

usama54321 committed:
const int should be parsed as uint64_t before casting

usama54321 commented:
Need to write tests for the initial commit + 4 bugfixes

usama54321 committed:
SILBuilder by default does not apply loc overrides for debug values, as
in the absence of a debug variable source loc, the instruction's loc
gets applied as the debug variable's loc which is incorrect. This patch
appropriately serializes the debug variable loc when possible and
reverts to the default behavior otherwise.

Also adds a test for debug_value serialization/deserialization

usama54321 committed:
Add test for disabled shadow copies in imported functions

usama54321 committed:
Add unreachable in readLoc

usama54321 commented:
@swift-ci test

usama54321 committed:
Print types of debug_value operand

usama54321 committed:
Add test to make sure generic signatures are serialized

usama54321 committed:
Add comment

usama54321 commented:
@swift-ci test

adrian-prantl commented on lib/IRGen/IRGenSIL.cpp:
```diff
@@ -965,6 +965,9 @@ class IRGenSILFunction :
   bool shouldShadowVariable(SILDebugVariable varInfo, bool isAnonymous) {
     return !IGM.IRGen.Opts.DisableDebuggerShadowCopies
       && !IGM.IRGen.Opts.shouldOptimize()
+      //Disable shadow in deserialized functions as they might be optimized
```
```suggestion
      // Shadow copies are only emitted at -Onone, but a deserialized function might have been already optimized, so ignore those.
```

adrian-prantl commented on lib/SIL/IR/SILBuilder.cpp:
```diff
@@ -638,17 +638,24 @@ DebugValueInst *SILBuilder::createDebugValue(SILLocation Loc, SILValue src,
+  SILDebugLocation DebugLoc;
+  if (overrideLoc) {
+    // Debug location overrides cannot apply to debug value instructions.
+    DebugLocOverrideRAII LocOverride{*this, std::nullopt};
```
Does putting this into the `{` block mean that it won't affect the new instruction? Is that what you want here?

usama54321 replied:
The loc override applied in the getSILDebugLocation below. I am keeping the old behavior consistent while allowing us to not override the loc in the else statement below while deserializing

adrian-prantl commented on lib/Serialization/SILFormat.h:
```diff
@@ -312,6 +314,19 @@ namespace sil_block {
+    
+    SILTypeCategoryField, // operand type category
+    SILTypeCategoryField, // debug var type category
+    BCFixed<11>,// poison, movableValueDebuginfo, trace, hasDebugVar, isLet, isDenseMapSingleton(two bits), hasSource, hasLoc, hasExpr
```
clang-format

usama54321 commented:
@swift-ci test please

usama54321 commented:
@swift-ci test

usama54321 committed:
Fix test

usama54321 committed:
Prevent optimization in test

usama54321 committed:
Sometimes the debug scope already attached to a SILFunction before
deserialization contains AST information which is needed in certain
cases. Do not overwrite the debug scope in that case

usama54321 committed:
Formatting and comments

usama54321 committed:
Increment module format

usama54321 commented:
@swift-ci test

adrian-prantl commented on lib/Serialization/DeserializeSIL.cpp:
```diff
@@ -430,9 +430,7 @@ SILDeserializer::readLoc(unsigned kind, SmallVectorImpl<uint64_t> &scratch) {
-
-  // LocationKind was not a recognized SILLocation::LocationKind.
-  return std::nullopt;
+  llvm_unreachable("Invalid LocationKind");
```
We should not use llvm_unreachable to deal with invalid input. Code like this can crash long-running tools like LLDB which makes for a very unpleasant user experience. Can you just return an error instead?

adrian-prantl commented on lib/Serialization/SerializeSIL.cpp:
```diff
@@ -974,7 +978,108 @@ void SILSerializer::writeSILInstruction(const SILInstruction &SI) {
+          ListOfValues.push_back(LC.second);
+          ListOfValues.push_back(FNameID);
+        } else if (RawLoc.isFilenameAndLocation()) {
+          // TODO: this is a workaround until rdar://problem/25225083 is
```
This comment isn't useful for Swift contributors without access to the link. Can you summarize what the problem is here?

adrian-prantl commented on lib/Serialization/SerializeSIL.cpp:
```diff
@@ -974,7 +978,108 @@ void SILSerializer::writeSILInstruction(const SILInstruction &SI) {
+    if (DebugVar) {
+      attrs |= 1 << 3; // has debug var
+      attrs |= DebugVar->isLet() << 4;
+      attrs |= DebugVar->isDenseMapSingleton << 5; // needs two bits
```
Can you adjust these comments to the LLVM/Swift coding style? (full sentences)

usama54321 committed:
Revert "Add unreachable in readLoc"

This reverts commit c1682c7526b5ae6b8450aa3d383024a541915efc.

usama54321 committed:
Fix comments

usama54321 commented:
@swift-ci test
